task 1/detect.cpp

```diff
@@ -0,0 +1,256 @@
+{
+    cv::Mat color_channel, src_light, src_dim;
+    std::vector<cv::Mat> channels;
+
+    cv::split(srcImage, channels);
+    color_channel = channels[ENEMY_COLOR]; // Enemy channel
+
+    cv::threshold(color_channel, src_light, light_threshold, 255, CV_THRESH_BINARY);
+    if (src_light.empty())
+        return false;
+    imagePreProcess(src_light);
+    // cv::imshow("Light", src_light);
+
+    cv::threshold(color_channel, src_dim, dim_threshold, 255, CV_THRESH_BINARY);
+    if (src_dim.empty())
+        return false;
+    imagePreProcess(src_dim);
+    // cv::imshow("Dim", src_dim);
+
+    std::vector<std::vector<cv::Point>> light_contours_light, light_contour_dim;
+    LightBars light_bars_light, light_bars_dim;
+    std::vector<cv::Vec4i> hierarchy_light, hierarchy_dim;
+    cv::findContours(src_light, light_contours_light, hierarchy_light, CV_RETR_CCOMP, CV_CHAIN_APPROX_NONE);
+    cv::findContours(src_dim, light_contour_dim, hierarchy_dim, CV_RETR_CCOMP, CV_CHAIN_APPROX_NONE);
+
+    for (int i = 0; i < light_contours_light.size(); i++)
+    {
+        if (hierarchy_light[i][2] == -1) // no parent element
+        {
+            cv::RotatedRect rect = cv::minAreaRect(light_contours_light[i]);
+            if (isValidLightBar(light_contours_light[i], rect))
+            {
+                light_bars_light.emplace_back(rect, Area_Ratio(light_contours_light[i], rect));
+            }
+        }
+    }
+
+    for (int i = 0; i < light_contour_dim.size(); i++)
+    {
+        if (hierarchy_dim[i][2] == -1) // no parent element
+        {
+            cv::RotatedRect rect = cv::minAreaRect(light_contour_dim[i]);
+            if (isValidLightBar(light_contour_dim[i], rect))
+            {
+                light_bars_dim.emplace_back(rect, Area_Ratio(light_contour_dim[i], rect));
+            }
+        }
+    }
+
+    std::vector<int> light_to_rm, dim_to_rm;
+    for (int l = 0; l < light_bars_light.size(); l++)
+    {
+        for (int d = 0; d < light_bars_dim.size(); d++)
+        {
+            if (isSameBar(light_bars_light[l], light_bars_dim[d]))
+            {
+                if (light_bars_light[l].area_ratio > light_bars_dim[d].area_ratio)
+                {
+                    dim_to_rm.emplace_back(d);
+                }
+                else
+                    light_to_rm.emplace_back(l);
+            }
+        }
+    }
+
+    sort(light_to_rm.begin(), light_to_rm.end(), [](int a, int b) { return a > b; });
+    sort(dim_to_rm.begin(), dim_to_rm.end(), [](int a, int b) { return a > b; });
+
+    for (auto i : light_to_rm)
+    {
+        light_bars_light.erase(light_bars_light.begin() + i);
+    }
+    for (auto i : dim_to_rm)
+    {
+        light_bars_dim.erase(light_bars_dim.begin() + i);
+    }
+    for (const auto &light : light_bars_light)
+    {
+        light_bars.emplace_back(light);
+    }
+    for (const auto &dim : light_bars_dim)
+    {
+        light_bars.emplace_back(dim);
+    }
+
```

Comment:
maybe you can add some explanations about this part.

Reply:
After reviewing the SJTU‘s codes, they claimed they yielded two Mats from a light and a dim thresholds to reduce the influence of environmental illumination, and then duplicate light bars were eliminated and redundant calculation was too, that is, the intersection of the two found light bars was taken.
But, as I am concerned, just using the light threshold will be better for my task, or that is my fault to understand